Geometry_Engine/Compute/OutlinesFromLines.cs

```diff
@@ -0,0 +1,427 @@
+                cluster.RemoveOutliers(distanceTolerance);
+                if (cluster.Count == 0)
+                    continue;
+
+                if (!cluster.IsCoplanar(distanceTolerance))
+                {
+                    BH.Engine.Base.Compute.RecordWarning("Some of the lines have been ignored in the process of creating outlines because they were not coplanar.");
+                    continue;
+                }
+
+                Output<Polyline, List<Line>> preprocessedLines = cluster.OuterAndInnerLines(distanceTolerance);
+                result.AddRange(OutlinesFromPreprocessedLines(preprocessedLines.Item1, preprocessedLines.Item2, distanceTolerance));
+            }
+
+            return result;
+        }
+
+
+        /***************************************************/
+        /****              Private Methods              ****/
+        /***************************************************/
+
+        private static List<Line> CleanUpAndSplitWithEachOther(this List<Line> lines, double distanceTolerance)
+        {
+            Plane fitPlane = lines.SelectMany(x => x.ControlPoints()).ToList().FitPlane();
+            if (fitPlane == null)
+                return null;
+
+            lines = lines.BooleanUnion(distanceTolerance, true);
+            List<Point> intersectingPoints = Query.LineIntersections(lines).CullDuplicates(distanceTolerance);
+            return lines.SelectMany(x => x.SplitAtPoints(intersectingPoints).Select(y => y.Project(fitPlane))).ToList().CullDuplicateLines(distanceTolerance);
+        }
+
+        /***************************************************/
+
+        private static List<List<Line>> Cluster(this List<Line> lines, double distanceTolerance)
+        {
+            double sqTol = distanceTolerance * distanceTolerance;
+            Func<Line, Line, bool> distanceFunction = (a, b) =>
+            a.Start.SquareDistance(b.Start) < sqTol ||
+            a.Start.SquareDistance(b.End) < sqTol ||
+            a.End.SquareDistance(b.Start) < sqTol ||
+            a.End.SquareDistance(b.End) < sqTol;
+            return lines.ClusterDBSCAN(distanceFunction);
+        }
+
+        /***************************************************/
+
+        private static void RemoveOutliers(this List<Line> lines, double distanceTolerance)
+        {
+            int initialCount = lines.Count;
+
+            Dictionary<Point, int> nodesByValence = new Dictionary<Point, int>();
+            foreach (Line l in lines)
+            {
+                if (!nodesByValence.ContainsKey(l.Start))
+                    nodesByValence.Add(l.Start, 0);
+
+                nodesByValence[l.Start]++;
+
+                if (!nodesByValence.ContainsKey(l.End))
+                    nodesByValence.Add(l.End, 0);
+
+                nodesByValence[l.End]++;
+            }
+
+            foreach (Point point in nodesByValence.Keys.ToList())
+            {
+                if (nodesByValence.ContainsKey(point))
```

Comment:
After the brief chat we had I removed this check without more thinking just to learn that it had been added on purpose 😃 It is needed because `CullValenceOne` method is recursive and may remove some keys from the `nodesByValence` dictionary. So actually it is an extremely important line of code.